sdk/src/main/scala/io/horizen/account/state/AccountStateView.scala

```diff
@@ -86,22 +87,59 @@ class AccountStateView(
 
   // after this we always reset the counters
   override def getFeePaymentsInfo(
-      withdrawalEpoch: Int,
-      consensusEpochNumber: ConsensusEpochNumber,
-      distributionCap: BigInteger,
-      blockToAppendFeeInfo: Option[AccountBlockFeeInfo] = None
+    withdrawalEpoch: Int,
+    consensusEpochNumber: ConsensusEpochNumber,
+    distributionCap: BigInteger,
+    blockToAppendFeeInfo: Option[AccountBlockFeeInfo] = None,
   ): (Seq[AccountPayment], BigInteger) = {
     var blockFeeInfoSeq = metadataStorageView.getFeePayments(withdrawalEpoch)
     blockToAppendFeeInfo.foreach(blockFeeInfo => blockFeeInfoSeq = blockFeeInfoSeq :+ blockFeeInfo)
     val mcForgerPoolRewards = getMcForgerPoolRewards(consensusEpochNumber, distributionCap)
     val poolBalanceDistributed = mcForgerPoolRewards.values.foldLeft(BigInteger.ZERO)((a, b) => a.add(b))
     metadataStorageView.updateMcForgerPoolRewards(mcForgerPoolRewards)
-    (AccountFeePaymentsUtils.getForgersRewards(blockFeeInfoSeq, mcForgerPoolRewards), poolBalanceDistributed)
+    if (Version1_4_0Fork.get(consensusEpochNumber).active) {
+      val (feePayments, delegatorPayments) =
+        getForgersAndDelegatorsShares(
+          AccountFeePaymentsUtils.getForgersRewards(blockFeeInfoSeq, mcForgerPoolRewards),
+          blockFeeInfoSeq,
+          mcForgerPoolRewards
+        )
+      metadataStorageView.updateForgerDelegatorPayments(delegatorPayments, consensusEpochNumber)
+
+      (feePayments ++ delegatorPayments.map(_.feePayment), poolBalanceDistributed)
+    } else {
+      (AccountFeePaymentsUtils.getForgersRewards(blockFeeInfoSeq, mcForgerPoolRewards), poolBalanceDistributed)
+    }
+  }
+
+  private[horizen] def getForgersAndDelegatorsShares(
+    feePayments: Seq[AccountPayment],
+    blockFeePaymentInfoSeq: Seq[AccountBlockFeeInfo],
+    mcForgerPoolRewards: Map[AddressProposition, BigInteger],
+  ): (Seq[AccountPayment], Seq[DelegatorFeePayment]) = {
+    val allPayments = feePayments.map { feePayment =>
+      // for each forger, get forger info from ForgerStakeV2 that contains reward share and reward address
+      val forgerInfo = blockFeePaymentInfoSeq
+        .find(fpi => fpi.forgerAddress == feePayment.address && fpi.blockSignPublicKey.isDefined && fpi.vrfPublicKey.isDefined)
+        .flatMap(fpi => getForgerInfo(ForgerPublicKeys(fpi.blockSignPublicKey.get, fpi.vrfPublicKey.get)))
+
+      // If forgerInfo present, calculate forger and delegators shares. If not, all goes to forger
+      forgerInfo match {
```

Comment:
please consider also that forgerInfo may be present and have:
rewardShare = 0
In this case all rewards goes to the forger, we could fall in the other case (to save some storage space)